crates/bevy_core_pipeline/src/taa/taa.wgsl

```diff
@@ -0,0 +1,183 @@
+// References:
+// https://www.elopezr.com/temporal-aa-and-the-quest-for-the-holy-trail
+// http://behindthepixels.io/assets/files/TemporalAA.pdf
+// http://leiy.cc/publications/TAA/TAA_EG2020_Talk.pdf
+// https://advances.realtimerendering.com/s2014/index.html#_HIGH-QUALITY_TEMPORAL_SUPERSAMPLING
+
+#import bevy_core_pipeline::fullscreen_vertex_shader
+
+#ifdef TONEMAP
+#import bevy_core_pipeline::tonemapping
+#endif
+
+@group(0) @binding(0) var view_target: texture_2d<f32>;
+@group(0) @binding(1) var history: texture_2d<f32>;
+@group(0) @binding(2) var velocity: texture_2d<f32>;
+@group(0) @binding(3) var depth: texture_depth_2d;
+@group(0) @binding(4) var nearest_sampler: sampler;
+@group(0) @binding(5) var linear_sampler: sampler;
+
+struct Output {
+    @location(0) view_target: vec4<f32>,
+    @location(1) history: vec4<f32>,
+};
+
+// The following 3 functions are from Playdead
+// https://github.com/playdeadgames/temporal/blob/master/Assets/Shaders/TemporalReprojection.shader
+fn RGB_to_YCoCg(rgb: vec3<f32>) -> vec3<f32> {
+    let y = (rgb.r / 4.0) + (rgb.g / 2.0) + (rgb.b / 4.0);
+    let co = (rgb.r / 2.0) - (rgb.b / 2.0);
+    let cg = (-rgb.r / 4.0) + (rgb.g / 2.0) - (rgb.b / 4.0);
+    return vec3(y, co, cg);
+}
+
+fn YCoCg_to_RGB(ycocg: vec3<f32>) -> vec3<f32> {
+    let r = ycocg.x + ycocg.y - ycocg.z;
+    let g = ycocg.x + ycocg.z;
+    let b = ycocg.x - ycocg.y - ycocg.z;
+    return saturate(vec3(r, g, b));
+}
+
+fn clip_towards_aabb_center(previous_color: vec3<f32>, current_color: vec3<f32>, aabb_min: vec3<f32>, aabb_max: vec3<f32>) -> vec3<f32> {
+    let p_clip = 0.5 * (aabb_max + aabb_min);
+    let e_clip = 0.5 * (aabb_max - aabb_min) + 0.00000001;
+    let v_clip = previous_color - p_clip;
+    let v_unit = v_clip / e_clip;
+    let a_unit = abs(v_unit);
+    let ma_unit = max(a_unit.x, max(a_unit.y, a_unit.z));
+    return select(previous_color, p_clip + v_clip / ma_unit, ma_unit > 1.0);
+}
+
+// TAA is ideally applied after tonemapping, but before post processing
+// Post processing wants to go before tonemapping, which conflicts
+// Solution: Put TAA before tonemapping, tonemap TAA input, apply TAA, invert-tonemap TAA output
+// https://advances.realtimerendering.com/s2014/index.html#_HIGH-QUALITY_TEMPORAL_SUPERSAMPLING, slide 20
+#ifdef TONEMAP
+fn rcp(x: f32) -> f32 { return 1.0 / x; }
+
+fn tonemap(color: vec3<f32>) -> vec3<f32> {
```

Comment:
It might be better to use https://gpuopen.com/learn/optimized-reversible-tonemapper-for-resolve/ for the reversible tonemapping (afaik it's faster and doesn't cause hue-shifting, though I'm not sure which variant you should use).

Reply:
I'm not sure which is better. I was playing around with this a few weeks ago and decided to use luma over max3. Which do you think looks better? Also note that they have different flicker patterns, but I can't capture that in a screenshot.

(No AA)
![image](https://user-images.githubusercontent.com/47158642/217322026-55f341af-aa95-4d90-9523-f61f007ae48e.png)

(TAA max3)
![image](https://user-images.githubusercontent.com/47158642/217322060-dc98f8e9-b640-4036-97d9-a96b0a4e4f74.png)

(TAA luma)
![image](https://user-images.githubusercontent.com/47158642/217322300-1121a23b-5c87-4c6d-bc7c-af1c44e92f6c.png)

Reply:
Decided to switch to max3.

Reply:
luma smears the colours in the cubes together quite a bit, while max3 doesn't.

Imo max3 looks better.

Reply:
The checkered cubes look noticably blurry to me on the `TAA luma` screenshot. `max3` looks better in my opinion (there's only a tiny bit of oversharpening).